draft-wang-ppm-dap-taskprov.md

```diff
@@ -222,144 +225,212 @@ struct {
+> this to the Author?
 
-Upon receiving a report, leader reads the extension codepoint in
-`extension_type`. If the leader does not support this extension, it MUST ignore
-it. In particular, if the task ID is not known, then it MUST abort the handshake
-with "unrecognizedTask".
+Once the client opts in to a task, it MAY begin uploading reports for the task.
+Each report MUST offer the `task_prov` extension with the `TaskConfig`
+disseminated by the Author as the extension payload. In addition, the report's
+task ID MUST be computed as described in {{construct-task-id}}.
 
-> See [#334](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/334) for
+# Leader Behavior
+
+## Upload Protocol
+
+Upon receiving a `Report` from the Client with the `task_prov` extension, if the
+Leader does not support the extension, it MUST ignore the extension payload and
+proceed as usual. In particular, if the task ID is not recognized, then it MUST
+abort the upload request with "unrecognizedTask".
+
+> NOTE: This behavior assumes unrecognized extensions are to be ignored. See
+> [#334](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/334) for
 > discussion.
 
-If aggregator supports `task-prov` extension, it should proceed to
-{{provisioning-a-task}}. If task provision failed, leader MUST alert the client
-with error from {{provisioning-a-task}}. If task provision succeeded, leader
-should continue to the rest of upload flow.
+Otherwise, if the Leader does support the extension, it first attemps to parse
+the payload. If parsing fails, it MUST abort with "unrecognizedMessage".
+
+Next, it checks that the task ID included in the report matches the task ID
+derived from the extension payload as specified in {{construct-task-id}}. If the
+task ID does not match, then the Leader MUST abort with "unrecognizedTask".
 
-Leader MAY return error to the client if task creation failed.
+The Leader then decides whether to opt in to the task as described in
+{{provisioning-a-task}}. If it opts out, it MUST abort the upload request with
+"invalidTask".
+
+> OPEN ISSUE: In case of opt-out, would it be useful to specify how to report
+> this to the Author?
+
+Finally, once the Leader has opted in to the task, it completes the upload
+request as usual.
+
+## Aggregate Protocol
+
+When the Leader opts in to a task, it derives the VDAF verification key for that
+task as described in {{vdaf-verify-key}}.
+
+## Collect Protocol
+
+The Collector might issue a collect request for a task provisioned by the
+`task_prov` extension prior to opting in to the task. In this case, the Leader
+would need to abort the collect request with "unrecognizedTask". When it does
+so, it SHOULD also include a "Retry-After" header in its HTTP response
+indicating the time after which the Collector should retry its request.
+
+> TODO: Find RFC reference for "Retry-After".
+
+> OPEN ISSUE: This semantics is awkward, as there's no way for the Leader to
```

Comment:
Which text? The intent is that each party should opt in or out, including the Collector.

Reply:
If a non-taskprov Collector sends a CollectReq and gets unrecognizedTask in response, then it's just supposed to quit, since it's expectation is that there's no point in the future where the Leader will be configured with the task. Instead, we're asking it to retry at some point in the future, but retrying is necessarily fruitless. Since Retry-After is typically handled "under the hood" by HTTP clients (in the reqwest crate for Rust, you have to explicitly override this), non-taskprov Collectors will appear to hang indefinitely here.

This can be avoided if the Collector has to explicitly opt-in.

Reply:
The leader can choose not to send retry-after header if it hasn't opt-in to task_prov, collector will just get unrecognizedTask and quit, the task_prov process happens inside leader/helper, so it should be sufficient for leader to decide whether a retry-after should be given to a collector.

Reply:
How does the collector know that the task is definitely never going to exist? Remember, the intention of this retry-after mechanism is to allow the Leader to keep the Collector on the phone in case the task gets configured in the near future. If a taskprov Leader can't distinguish whether the Collector is taskprov or non-taskprov, it either has to ignore this SHOULD or risk keeping a non-taskprov Collector in an indefinite retry-after loop.

TBH, I would be happy to avoid this complexity by dropping the retry-after behavior altogether, at least until we can implement it unambiguously.